pytorch_lightning/plugins/precision/double.py

```diff
@@ -0,0 +1,63 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from functools import wraps, partial
+from typing import Any, Callable, Sequence, Tuple, TYPE_CHECKING
+
+import torch
+
+from pytorch_lightning.plugins.precision.precision_plugin import PrecisionPlugin
+from pytorch_lightning.utilities.apply_func import apply_to_collection
+
+if TYPE_CHECKING:
+    from torch.nn import Module
+    from torch.optim import Optimizer
+
+
+class DoublePrecisionPlugin(PrecisionPlugin):
+    """Plugin for training with double (`torch.float64`) precision."""
+
+    precision: int = 64
+
+    @staticmethod
+    def _to_double_precision(tensor: torch.Tensor) -> torch.Tensor:
+        if tensor.is_floating_point():
+            return tensor.to(dtype=torch.float64)
+        return tensor
+
+    @staticmethod
+    def _teardown(model: 'Module', old_forward: Callable) -> None:
+        model.forward = old_forward
+
+    def _patch_forward(self, model: 'Module') -> None:
+        old_forward = model.forward
+
+        @wraps(old_forward)
+        def new_forward(*args, **kwargs) -> Any:
+            return old_forward(*apply_to_collection(args, torch.Tensor, DoublePrecisionPlugin._to_double_precision),
+                               **apply_to_collection(kwargs, torch.Tensor, DoublePrecisionPlugin._to_double_precision))
```

Comment:
You can apply the same pattern as here:

https://github.com/PyTorchLightning/pytorch-lightning/blob/3a56a6024e0d8b239801cce558381807c24ba3d0/pytorch_lightning/plugins/training_type/deepspeed.py#L52-L58

By checking for `torch.DoubleTensor`

Reply:
The method I used allowed for filtering using `torch.is_floating_point`, which is more reliable than just type checking for `FloatTensor` and `cuda.FloatTensor`. I can change this if needed but I think `is_floating_point` is a better approach. I modified this from the source for Nvidia Apex, which seems quite reliable :)